### ISSUE TITLE
New @RootComponent annotation to Beans

### ISSUE THREAD

**Issue by smaugho** [androidannotations/androidannotations#2198]
Hi, I would like to add to the annotation @RootComponent. The idea is to inject the component which injected the Bean (or Bean tree). 

There are several scenarios when this is needed. For instance, there are many libraries (ie. Facebook Login) which have an initialization method, which takes a parameter "Fragment, Activity" depending of where it is being initialized. If you implement a Bean to hold the logic to interact with that library (Let's say, some FacebookHelper, or Controller... or SocialManager), you'll need to know if the Bean was initialized in an Activity or Fragment. 

Let's say that the Bean is injected by other (ie. a ViewModel EBean), then it is not even in contact with the Activity which creates it.. 

A see a very simple implementation for the feature.. in the getInstance_ of a Bean, we provide not only the context, but also the rootComponent (if any).. if this Bean injects others, it passes the component as well. 

The annotation @RootComponent would simply expose that property, similar to how @RootContext work. 

I really was thinking in naming the annoation @RootView, but I think is much more generic, since it could work in other "generating components" .. though maybe there is a better name for it. 

Please let me know it if fine to add it, I have already an implementation around, would be simply to insert it in AndroidAnnotations. 

As a remark, this doesn't apply to the Singleton scope of the beans, but there are different scenarios in which it could prove to be beneficial.

**Comment by smaugho** [comment]
Note, there is another benefit of doing this. It would permit to easily implement #1488 . 

Basically for implementing the Activity Scope, we could save a static Map with keys "String or number" which references created beans in the Bean. When it is being created, it can check the rootComponent variable passed in the getInstance_, in the case of Activity scope, it checks if this rootComponent is the fragment or activity, if it exists in the saved instances, then it uses that one (so, the Bean will be only one per activity). 

Note please, that I said "String or number", probably we could save some unique identifier to the activity, but clearly we cannot save in a Static variable a reference to the activity itself (that would generate a big leak to the Activity object).

**Comment by WonderCsabo** [comment]
@smaugho i am not sure it can be implemented. You can inject a bean into any kind of class. To do `@RootComponent`, you want to actually restrict the bean to just be injected to one specific class. That is not acceptable.

**Comment by smaugho** [comment]
Well, the idea of `@RootComponent` would be to provide the "initial component" which injected the Bean (not just certain type of classes, I wanted to use it for only views-related, and that's why I thought in `@RootView`, but that will limit the functionality, so the idea of `@RootComponent` is more generic). 

The logic to determine the `rootComponent` could be completely similar to the logic used for the context. So the `EComponentHolder` implements `IJExpression getRootComponentRef()`, which is `_this()` by default. 

`EBeanHolder` overrides the method and returns the field "rootComponent_", which is set in the same way that `context_` is set for the bean. 

When a `@Bean` is injected, the `getInstance_`  method is called with both parameters which are read from the `holder.getRootComponentRef()`

--- 

About implementing **scope=Activity** as requested in #1488 , lets say the user injects a Bean marked with that scope in a place not Activity dependent (ie. `@EService` or `@EApplication`). I think it is completely ok to throw a runtime exception (or emit a warning depending of what we determine the best option), something like that the user is injecting the Bean outside of an activity Scope. So simply, we check the `rootComponent` passed in the `getInstance_`, and only Views/Activities/Fragment are accepted as valid in that kind of Bean (checked with a simple instance of). AA already emits these kind of warning, for instance if the user makes:

```
@RootContext
Activity activity;
```
if the context is not an activity, a warning is sent and "activity" is not injected. 

When the Bean is injected in another view, fragment or activity, we can always check if it is in the Activity Context. I mentioned above to save the Bean in a static Map.. but to be honest sounds quite a bad solution, cause we will not know when we should remove the Bean from the map for being removed from memory. 

So I thought a solution for it, let's see how you see it. But it should work, and it is fine with the architecture. 

We make the Activity implement an interface with the method:

```
public interface IBeansFactory {
   void getBean(String beanId);
}
```

So that the app manages its own list of beans. The `getInstance_` will call that in the activity which is passed as rootView, so the Bean is injected (in the activity context), and once that the activity is destroys, it releases its list of beans. 

To be honest, seeing it in this way, the Context typically points to the activity inside Fragments/Views (so maybe we could implement this with only the Context in fact). Though the requirements of having the fragment which injected the view is not met (for things like the explained in the beginning)

**Comment by WonderCsabo** [comment]
@smaugho can you post example generated code for the `@RootComponent` functionality?

**Comment by smaugho** [comment]
Sure @WonderCsabo .. firstly for the bean, let's say the bean is injected in a @EFragment, the injection would be something like:

```java
//Declared in annotated EFragment
@Bean
MyBean mBean;

//Generated code
private void init_() {
    this.mBean = MyBean_.getInstance_(getContext(), this);
}
```

Code inside the bean:

```java
private Context context_;
private Object rootComponent_;

private MyBean_(Context context, Object rootComponent) {
        context_ = context;
        rootComponent_ = rootComponent;
        init_();
}

public static MyBean_ getInstance_(Context context, Object rootComponent) {
        return new MyBean_(context, rootComponent);
}
```

Injection of a Bean inside another bean:

```java
//Declared in annotated EBean
@Bean
MyBean mBean;

//Generated code
private void init_() {
    this.mBean = MyBean_.getInstance_(getContext(), rootComponent_);
}
```

Injection with `@RootComponent`

```java
//Declared in a Bean
@RootComponent
Fragment mFragment;

//Generated code
private void init_() {
        if (rootComponent_instanceof Fragment) {
            this.mFragment = ((Fragment) rootComponent_);
        } else {
            Log.w("MyBean_", (("Due to Root Comonent class "+ rootComponent_.getClass().getSimpleName())+", the @RootComponent Fragment won't be populated"));
        }
}
```

Use case example (Helper for Facebook SDK login)

```java
@EBean
public class FacebookLoginHelper {

    @RootComponent
    Object rootView;

    //.... other declarations/injections
   
    public void login(FacebookCallback<LoginResult> loginResult) {

        if (rootView instanceof Activity) {
            loginManager.logInWithReadPermissions((Activity)rootView, permissions);
        } else if (rootView instanceof Fragment) {
            loginManager.logInWithReadPermissions((Fragment) rootView, permissions);
        } else if (rootView instanceof android.support.v4.app.Fragment) {
            loginManager.logInWithReadPermissions((android.support.v4.app.Fragment) rootView, permissions);
        }

        loginManager.registerCallback(callbackManager, loginResult);
    }


    public void facebookLoginOnActivityResult(int requestCode, int resultCode, Intent data) {
        if (callbackManager != null) {
            callbackManager.onActivityResult(requestCode, resultCode, data);
        }
    }

}

```

---------

About the proposed solution for adding `scope = Scope.Activity` to Beans I'll add it to the issue #1488 .

**Comment by smaugho** [comment]
About the `@ComponentRoot`, I think that most of the issues it will solve, would be resolved if #1636 which adds `@RootFragment` is implemented. So, I would work on that specific issue and I'll submit an implementation. I'll close this issue then ;).
